### PR TITLE
⚡️ Calculate positive and negative FMA sums in parallel

### PR DESCRIPTION
**Pull request by michael-platzer**
The FMA currently calculates the sum of the product and the addend first and, in case certain conditions are met, it then negates it.

https://github.com/openhwgroup/cvfpu/blob/eba1bbebd84121a529a4696c5ce792da3013a8bd/src/fpnew_fma_multi.sv#L474-L475

However, two's complement negation requires another full carry chain over the width of the sum, which means that the FMA currently has two sequential carry chains over the entire `3*PRECISION_BITS+4` bits of the sum (one for the addition and another one for the negation). This leads to long paths and timing issues.

A common way to reduce the length of signal paths in a FP adder is to calculate a positive and a negative sum in parallel and then select the correct result with a mux. The advantage is that the carry chains are no longer sequential but parallel instead. The Wally core uses this technique in its FMA: https://github.com/openhwgroup/cvw/blob/main/src/fpu/fma/fmaadd.sv

This PR modifies the FMA of CVFPU accordingly. We have verified that it is equivalent to the current implementation with Synopsys VC formal's sequential equivalence check (the script to reproduce the check is posted in a comment below).

Initial results from preliminary PD runs in FC suggest that TNS is reduced and timing ameliorated.

### CODE REVIEW

**Comment by michael-platzer** [comment]
The following script can be used to verify that the proposed changes are sequentially equivalent to the current implementation with Synopsys VC formal 's sequential equivalence check (start VC formal via `vcf -fmode SEQ`, then `source` this script):

```
set_fml_appmode SEQ

set flist_golden [list \
 "common_cells/lzc.sv" \
 "cvfpu/src/fpnew_pkg.sv" \
 "cvfpu/src/fpnew_classifier.sv" \
 "cvfpu/src/fpnew_rounding.sv" \
 "cvfpu/src/fpnew_fma_multi.sv" \
]
set flist_impl [list \
 "common_cells/lzc.sv" \
 "cvfpu/src/fpnew_pkg.sv" \
 "cvfpu/src/fpnew_classifier.sv" \
 "cvfpu/src/fpnew_rounding.sv" \
 "cvfpu/src/fpnew_fma_multi_new.sv" \
]

analyze -format sverilog -library spec -vcs $flist_golden +incdir+.
analyze -format sverilog -library impl -vcs $flist_impl +incdir+.

elaborate_seq -spectop fpnew_fma_multi -impltop fpnew_fma_multi

map_by_name -clock spec.clk_i

create_clock -period 100 spec.clk_i
create_reset spec.rst_ni -sense low

fvassume -expr {spec.src_fmt_i == 0}
fvassume -expr {spec.src2_fmt_i == 0}
fvassume -expr {spec.dst_fmt_i == 0}

sim_run -stable
sim_set_state -uninitialized -apply 0

check_fv -block

report_proofs
```

Note that the `+incdir+` needs to point to a path where the `common_cells/registers.svh` header can be found. `cvfpu/src/fpnew_fma_multi_new.sv` contains the changes of this patch, while `cvfpu/src/fpnew_fma_multi.sv` and all other source files are the current version from `develop`. Different source and destination formats can be tried manually (unfortunately, runtime explodes when attempting to constrain these more loosely via, e.g., `spec.src_fmt_i inside {0,1,2,3,4}`).

We have run the script for various source and destination formats and proved that the changes are equivalent:

```
  Summary Proofs:
   ----------------------------------------------------------------------------------------------------------------------
    VpId |           Name |      Type |         Parent |     #A |     #C |     #S |     #F |     #I |    Status |     %
   ----------------------------------------------------------------------------------------------------------------------
       0 |         seqdef |      root |            nil |     13 |      5 |     13 |      0 |      0 |   success |   100
       0 |      seqdef-rw |        or |         seqdef |      - |      - |      - |      - |      - |         - |     -
       0 |          rw1_1 |       int |      seqdef-rw |      5 |      5 |      5 |      0 |      0 |   success |   100
       0 |       rw1_1-ur |        or |          rw1_1 |      - |      - |      - |      - |      - |         - |     -
       0 |           ur_1 |      leaf |       rw1_1-ur |      4 |      5 |      4 |      0 |      0 |   success |   100
       0 |      rw1_1-dcp | decompose |          rw1_1 |      - |      - |      - |      - |      - |         - |     -
       0 |         idcp_1 |      leaf |      rw1_1-dcp |      4 |      5 |      4 |      0 |      0 |   success |   100
   ----------------------------------------------------------------------------------------------------------------------
```

**Review by stmach — APPROVED** [review]
LGTM thank you very much!

A thing to consider would be whether this architectural change should be behind a parameter as it trades area for propagation delay, but I'll leave that up to the maintainers!

**Comment by michael-platzer** [comment]
Thanks a lot for the review!

> A thing to consider would be whether this architectural change should be behind a parameter as it trades area for propagation delay, but I'll leave that up to the maintainers!

Well, actually, in our trials the area was slightly reduced by this change but that is probably a side effect of the improved timing.

**Comment by zarubaf** [comment]
@lucabertaccini do you want to have a parameter or are we good to merge "as is"?

**Comment by lucabertaccini** [comment]
Thanks a lot @michael-platzer! This is great :) Considering that the area gets slightly better as a side effect of the reduced timing pressure, I think we can merge it "as is" and avoid adding an extra parameter.

**Review by davideschiavone — APPROVED** [review]

**Pull request merged**
